data_structures/avltree_map.hpp

```diff
@@ -0,0 +1,957 @@
+/**
+ * @file
+ * @brief An [AVLTree](https://en.wikipedia.org/wiki/AVL_tree)-based
+ *  map implementation
+ * @details The map is sorted according to the natural ordering of its
+ *  keys or by a {@code Compare} function provided; This implementation
+ *  provides guaranteed log(n) time cost for the contains, get, insert
+ *  and remove operations.
+ * @author [r.ivance](https://github.com/RIvance)
+ */
+
+#ifndef AVLTREE_MAP_HPP
+#define AVLTREE_MAP_HPP
+
+#include <cstdint>      /// for standard integer definitions
+#include <cstddef>      /// for standard definitions like `std::size_t`
+#include <cassert>      /// for assert
+#include <memory>       /// for smart pointers
+#include <functional>   /// for functional programming
+#include <stack>        /// for `std::stack`
+#include <vector>       /// for `std::vector`
+
+/**
+ * An AVLTree-based map implementation
+ * https://en.wikipedia.org/wiki/AVL_tree
+ * @tparam Key the type of keys maintained by this map
+ * @tparam Value the type of mapped values
+ * @tparam Compare
+ */
+template <
+    typename Key, typename Value,
+    typename Compare = std::less<Key>
+>
+class AvlTreeMap
+{
+  private:
+
+    using USize = size_t;
+    using Factor = int64_t;
```

Comment:
Please add a one-line description of what the variable is for (**see the example below**; check other parts of the code as well).

```c++
int64_t data = 0;       ///< The value/key of the node.
```